code/modules/organs/internal/_internal.dm

```diff
@@ -3,96 +3,80 @@
+	. = ..()
+	if(species.organs_icon)
 		icon = species.organs_icon
 
 //disconnected the organ from it's owner but does not remove it, instead it becomes an implant that can be removed with implant surgery
 //TODO move this to organ/internal once the FPB port comes through
 /obj/item/organ/proc/cut_away(var/mob/living/user)
 	var/obj/item/organ/external/parent = owner.get_organ(parent_organ)
 	if(istype(parent)) //TODO ensure that we don't have to check this.
-		removed(user, 0)
+		uninstall(detach = TRUE)
 		LAZYADD(parent.implants, src)
 
-/obj/item/organ/internal/removed(var/mob/living/user, var/drop_organ=1, var/detach=1)
-	if(owner)
-		owner.internal_organs -= src
-		owner.internal_organs_by_name -= organ_tag
-		var/obj/item/organ/external/E = owner.get_organ(parent_organ)
-		E.update_internal_organs_cost()
-
-		if(detach)
-			var/obj/item/organ/external/affected = owner.get_organ(parent_organ)
-			if(affected)
-				LAZYREMOVE(affected.internal_organs, src)
-				status |= ORGAN_CUT_AWAY
-	..()
-
-/obj/item/organ/internal/replaced(var/mob/living/carbon/human/target, var/obj/item/organ/external/affected)
-
-	if(!istype(target))
-		return 0
-
+/obj/item/organ/internal/install(mob/living/carbon/human/target, obj/item/organ/external/affected, in_place)
 	if(status & ORGAN_CUT_AWAY)
-		return 0 //organs don't work very well in the body when they aren't properly attached
+		return
 
+	//#FIXME: This feels like a hack
+		//# Also it probably should be of the species of the parent limb tbh..
 	// robotic organs emulate behavior of the equivalent flesh organ of the species
-	if(BP_IS_PROSTHETIC(src) || !species)
-		species = target.species
+	// if(BP_IS_PROSTHETIC(src) || !species)
+	// 	set_species(target.species)
 
-	..()
+	if(!(. = ..()))
+		return
 
 	STOP_PROCESSING(SSobj, src)
-	target.internal_organs |= src
-	target.internal_organs_by_name[organ_tag] = src
 	LAZYDISTINCTADD(affected.internal_organs, src)
-	return 1
+	affected.cavity_max_w_class = max(affected.cavity_max_w_class, w_class)
+	affected.update_internal_organs_cost()
 
-/obj/item/organ/internal/die()
-	..()
-	if((status & ORGAN_DEAD) && dead_icon)
-		icon_state = dead_icon
+/obj/item/organ/internal/uninstall(in_place, detach, ignore_children)
+	//Make sure we're removed from whatever parent organ we have, either in a mob or not
+	var/obj/item/organ/external/P = null
+	if(owner)
+		P = owner.get_organ(parent_organ)
+	else if(istype(loc, /obj/item/organ/external))
+		var/obj/item/organ/external/E = loc
+		if(E.organ_tag == parent_organ)
+			P = E
+
+	if(P)
+		LAZYREMOVE(P.internal_organs, src)
+		P.update_internal_organs_cost()
+	. = ..()
 
 /obj/item/organ/internal/remove_rejuv()
-	if(owner)
-		owner.internal_organs -= src
-		owner.internal_organs_by_name -= organ_tag
-		while(null in owner.internal_organs)
-			owner.internal_organs -= null
-		var/obj/item/organ/external/E = owner.organs_by_name[parent_organ]
-		if(istype(E))
-			LAZYREMOVE(E.internal_organs, src)
+	uninstall()
```

Comment:
Is this kind of thing legal with the new structure? Don't you have to call the mob's removal proc to update the lists in the block I commented on above?

Reply:
Well, the idea is that uninstall/install is self-contained. And since when we rejuv we actually just reset the organs (since I added a check to avoid having to recreate all organs and just carpet delete everthing) we only care that the organ itself is in an uninstalled state. Before "remove_rejuv" was just "removed()" but without uninstallation effects and clearing the mob refs so it could get deleted by create_organs properly. But now create_organs doesn't just mass delete everything so..

Its not necessarily great, but I've been trying to plug in my changes to get things to work, and assuming we could always go back and fix this kind of things once we get around to fixing more stuff.

Rejuvination is still doing some hacky stuff to avoid organs breaking everything. It could always be done in other ways if we can work out rejuvination too.